### ISSUE TITLE
fatal error using gpuInfo()

### ISSUE THREAD

**Issue by pathfinder11** [cdeterman/gpuR#27]
I have installed the gpuR packge on my windows machine without any issues(no compilation or install errors). However when I go to run some basic commands from the packge, like gpuInfo(), detectGPUs(), detectPlatforms() etc... R studio seems to take a long time to respond and eventually comes up with a fatal error from R and the program closes. Do you have any suggestions as to why this might happen? Is it a user permissions thing?
Thanks in advance.

**Comment by cdeterman** [comment]
Hi @pathfinder11, I assume you installed the package from CRAN?  What GPU and SDK (NVIDIA, AMD, Intel, etc.) do you have installed?

Have you tried compiling any C++/OpenCL code previously to confirm your SDK is installed properly and functioning?  If not, you can use the C++ file I have [here](https://github.com/cdeterman/intel_opencl/blob/master/clDeviceQuery.cpp) and compile it with the following call (substituting appropriate paths of course) and assuming you have Rtools installed in the default location.

```sh
C:/Rtools/mingw_64/bin/g++ clDeviceQuery.cpp -o clDeviceQuery -I"C:\path\to\SDK\include" -L"C:\Path\to\SDK\lib\x64" -lOpenCL
./clDeviceQuery.exe
```

Note that this refers to the 64 bit versions so make sure everything is pointing to the appropriate location (hence the `x64` example depending upon SDK).

**Comment by pathfinder11** [comment]
Hi @cdeterman, yes the package was downloaded from CRAN. I am using NVIDIA GPU at present (NVS 315), just as a starting point. I have checked that it is compatible with CUDA. For some reason I can't get this C++ file to compile.

`C:\Rtools\gcc-4.6.3\bin>g++ clDeviceQuery.cpp -o clDeviceQuery -I"C:/Program~Files/NVIDIA~GPU~Computing~Toolkit/CUDA/v7.5/include" -L"C:/Program~Files/NVIDIA~GPU~Computing~Toolkit/CUDA/v7.5/lib/x64" -lOpenCL
clDeviceQuery.cpp:8:19: fatal error: CL/cl.h: No such file or directory
compilation terminated.
C:\Rtools\gcc-4.6.3\bin>`

Similarly can seem to compile CUDA example files. I can't seem to explain why the compiler can't find these header files, even though they are in the included paths.

**Comment by cdeterman** [comment]
@pathfinder11, hmm... that is odd.  Are you certain following the path on your machine as the `include/CL` directory?  Also I see the `~` symbols in your paths.  I don't think those are usually there in Windows systems.  A space is acceptable with the quotes so if there are spaces in your paths you may leave them there.

**Comment by pathfinder11** [comment]
Hi @cdeterman, the CL directory is definitely there. If I take out the tilde sysmbol and leave the space in the directory paths I get a different problem to do with the OpenCL.lib file..

`C:\Rtools\gcc-4.6.3\bin>g++ clDeviceQuery.cpp -o clDeviceQuery -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v7.5\include" -L"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v7.5\lib\x64" -lOpenCL
c:/rtools/gcc-4.6.3/bin/../lib/gcc/i686-w64-mingw32/4.6.3/../../../../i686-w64-mingw32/bin/ld.exe: skipping incompatible C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v7.5\lib\x64/OpenCL.lib when searching for -lOpenCL
c:/rtools/gcc-4.6.3/bin/../lib/gcc/i686-w64-mingw32/4.6.3/../../../../i686-w64-mingw32/bin/ld.exe: skipping incompatible C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v7.5\lib\x64/OpenCL.lib when searching for -lOpenCL
c:/rtools/gcc-4.6.3/bin/../lib/gcc/i686-w64-mingw32/4.6.3/../../../../i686-w64-mingw32/bin/ld.exe: skipping incompatible C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v7.5\lib\x64\OpenCL.lib when searching for -lOpenCL
c:/rtools/gcc-4.6.3/bin/../lib/gcc/i686-w64-mingw32/4.6.3/../../../../i686-w64-mingw32/bin/ld.exe: cannot find -lOpenCL
collect2: ld returned 1 exit status`

**Comment by cdeterman** [comment]
@pathfinder11 that looks like the `g++` is defaulting to the 32-bit version.  If you pass the `-L` option as `-L"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v7.5\lib\Win32"` it should compile.

If you want to do the 64 bit version you will need to call the correct `g++` explicitly.  You should have a `mingw_64` directory within `C:\Rtools`.  In side there would be the 64 bit version.  You would want to use the path to there as I provided above.

**Comment by pathfinder11** [comment]
OK this part of the problem is solved. I had to install a later version of Rtools (Rtools34.exe) in order to get the C:\Rtools\mingw_64\bin directory  and then edit the PATH to include the C:\RTools\mingw_64\bin path.
Now back to the original issue. I still have the problem of  fatal error in R when I run something like gpuInfo().

**Comment by cdeterman** [comment]
@pathfinder11 does the compiled `clDeviceQuery` object run without problem though?  What is the output?  Does it actually find your GPU?

**Comment by pathfinder11** [comment]
Sure, the compiled executable of clDeviceQuery is all good, GPU device found. It's just the operation within of the toolbox commands within R that appears to be the problem. Given that I have updated Rtools I decided to try to reinstall the gpuR pacakge, but now I am getting errors/issues with the install. These issues seem related to another open issue by @skanskan , so will post further questions under this issue to avoid duplication.
